python/ray/test/cluster_utils.py

```diff
@@ -0,0 +1,181 @@
+                passed in.
+            head_node_args (kwargs): Arguments to be passed into
+                `start_ray_head` via `self.add_node`.
+        """
+        self.head_node = None
+        self.worker_nodes = {}
+        self.redis_address = ""
+        if initialize_head:
+            self.add_node(**head_node_args)
+            if connect:
+                ray.init(redis_address=self.redis_address)
+        elif connect:
+            logger.warning("Head not started; not connecting.")
+
+    def add_node(self, **kwargs):
+        """Adds a node to the local Ray Cluster.
+
+        Arguments:
+            kwargs: Keyword arguments used in `start_ray_head`
+                and `start_ray_node`.
+
+        Returns:
+            Node object of the added Ray node.
+        """
+        if self.head_node is None:
+            address_info = services.start_ray_head(
+                node_ip_address=services.get_node_ip_address(),
+                include_webui=False,
+                cleanup=True,
+                use_raylet=True,
+                **kwargs)
+            self.redis_address = address_info["redis_address"]
+            # TODO(rliaw): Find a more stable way than modifying global state.
+            process_dict_copy = services.all_processes.copy()
+            for key in services.all_processes:
+                services.all_processes[key] = []
+            node = Node(process_dict_copy)
+            self.head_node = node
+        else:
+            address_info = services.start_ray_node(
+                services.get_node_ip_address(),
+                self.redis_address,
+                cleanup=True,
+                use_raylet=True,
+                **kwargs)
+            # TODO(rliaw): Find a more stable way than modifying global state.
+            process_dict_copy = services.all_processes.copy()
+            for key in services.all_processes:
+                services.all_processes[key] = []
+            node = Node(process_dict_copy)
+            self.worker_nodes[node] = address_info
+        logging.info("Starting Node with raylet socket {}".format(
+            address_info["raylet_socket_names"]))
+
+        return node
+
+    def remove_node(self, node):
+        """Kills all processes associated with worker node.
+
+        Args:
+            node (Node): Worker node of which all associated processes
+                will be removed.
+        """
+        if self.head_node == node:
+            self.head_node.kill_allprocesses()
+            self.head_node = None
+            # TODO(rliaw): Do we need to kill all worker processes?
+        else:
+            node.kill_allprocesses()
+            self.worker_nodes.pop(node)
+
+        assert not node.any_processes_alive(), (
+            "There are zombie processes left over after killing...")
+
+    def wait_for_nodes(self):
+        """Waits for all nodes to be registered with global state."""
+        try:
```

Comment:
What's the point of this `try/except`?

Reply:
If you want to avoid a bad error message when this is called before `ray.init`, you can check `ray.is_connected`